lib/charms/kratos_external_idp_integrator/v0/kratos_external_provider.py

```diff
@@ -0,0 +1,347 @@
+        self.redirect_uri = redirect_uri
+
+    def snapshot(self):
+        """Save redirect_uri."""
+        return {"redirect_uri": self.redirect_uri}
+
+    def restore(self, snapshot):
+        """Restore redirect_uri."""
+        self.redirect_uri = snapshot["redirect_uri"]
+
+
+class InvalidClientConfigEvent(EventBase):
+    """Event to notify the charm that the requirer's databag is invalid."""
+
+    def __init__(self, handle, error):
+        super().__init__(handle)
+        self.error = error
+
+    def snapshot(self):
+        """Save error."""
+        return {"error": self.error}
+
+    def restore(self, snapshot):
+        """Restore error."""
+        self.error = snapshot["error"]
+
+
+class ExternalIdpProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `ExternalIdpProvider`."""
+
+    redirect_uri_changed = EventSource(RedirectURIChangedEvent)
+    invalid_client_config = EventSource(InvalidClientConfigEvent)
+
+
+class ExternalIdpProvider(Object):
+    """Forward client configurations to Identity Broker."""
+
+    on = ExternalIdpProviderEvents()
+
+    def __init__(self, charm, client_config, relation_name=DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.update_client_config(client_config)
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self._on_provider_endpoint_relation_joined)
+        self.framework.observe(
+            events.relation_changed, self._on_provider_endpoint_relation_changed
+        )
+        self.framework.observe(
+            events.relation_departed, self._on_provider_endpoint_relation_departed
+        )
+
+    def _on_provider_endpoint_relation_joined(self, event):
+        if not self._client_config:
+            return
+
+        self._set_client_config(**self._client_config)
+
+    def _on_provider_endpoint_relation_changed(self, event):
+        data = event.relation.data[event.app]
+        if len(data) == 0:
+            return
+        redirect_uri = data[0].get("redirect_uri")
+        self.on.redirect_uri_changed.emit(redirect_uri=redirect_uri)
+
+    def _on_provider_endpoint_relation_departed(self, event):
+        self.on.redirect_uri_changed.emit(redirect_uri="")
+
+    def create_client(self):
+        """Parse the charm config and pass it to the relation databag."""
+        return self._set_client_config()
+
+    def remove_client(self):
+        """Remove the client config to the relation databag."""
+        # Do we need to iterate on the relations? There should never be more
+        # than one
+        for relation in self._charm.model.relations[self._relation_name]:
+            relation.data[self._charm.app].clear()
+
+    def update_client_config(self, config):
+        """Update the client config."""
+        try:
+            self._client_config = self._handle_config(config)
```

Comment:
I tried to separate configuring the Provider class and creating the client, when the user configures the class (calls `update_client_config`) then the config is validated, when the user creates the client using the config (calls `create_client`) then there is no need to validate the config because everytime the config is set we validate it.

Maybe the naming is not very clear. Do you think I should rework this logic?

Reply:
`create_client` took the config from `update_client_config`, so the config had to already be validated. Having said that, you are right that it was a little complicated for no reason.

I removed `update_client_config` and tried to simplify the logic. Is it any better now?